### PR TITLE
CollectBaseDistributionByCycle in Spark + tests

### LINKED ISSUE #397: implement CollectBaseDistributionByCycle in Spark

### LINKED ISSUE #1000: CollectBaseDistributionByCycle in Spark
implement CollectBaseDistributionByCycle in spark and match the results from Picard.

should be straightforward - mimic the structure from MeanQualityByCycleSpark

### PR DESCRIPTION
**Pull request by akiezun**
fixes #1000 

for @davidadamsphd

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+@CommandLineProgramProperties(
+        summary = "Program to chart the nucleotide distribution per cycle in a SAM or BAM file.",
+        oneLineSummary = "Program to chart the nucleotide distribution per cycle in a SAM or BAM file.",
+        programGroup = SparkProgramGroup.class
+)
+public final class CollectBaseDistributionByCycleSpark extends GATKSparkTool {
+
+    private static final long serialVersionUID = 1L;
+
+    @Argument(doc = "uri for the output file: a local file path",
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: Create a ticket to make this work on HDFS and add a comment here.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        private int maxLengthSoFar = 0;
+        private final long[][] firstReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] firstReadCountsByCycle = new long[maxLengthSoFar];
+        private final long[][] secondReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] secondReadCountsByCycle = new long[maxLengthSoFar];
+        private boolean seenSecondEnd = false;
+
+        /**
+         * Merges the argument into this histogram generator. Returns the modified 'this' object.
+         * Note: you can only merge HistogramGenerator is they have the same 'useOriginalQualities' value.
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: Please fix comment.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                case 'C':
+                case 'c':
+                    return 1;
+                case 'G':
+                case 'g':
+                    return 2;
+                case 'T':
+                case 't':
+                    return 3;
+            }
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: Add an explicit check for 'U' 'u' and throw if none of the real options were found.

**Comment by davidadamsphd** [comment]
I started reviewing, but wasn't able to finish tonight. I'll continue tomorrow.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    private static final class HistogramGenerator implements Serializable{
+        private static final long serialVersionUID = 1L;
+
+        private static final int NBASES = 5; //ACGT + unknown
+        private int maxLengthSoFar = 0;
+        private final long[][] firstReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] firstReadCountsByCycle = new long[maxLengthSoFar];
+        private final long[][] secondReadTotalsByCycle = new long[NBASES][maxLengthSoFar];
+        private long[] secondReadCountsByCycle = new long[maxLengthSoFar];
+        private boolean seenSecondEnd = false;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: I found this variable name a little confusing, I think seenSecondOfPair would be clearer (or just add a comment).

Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: or maybe seenAnySecondOfPair (because this appears to only be used as a signal that there is any data from the second read of pairs).

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            if (bases == null) {
+                return this;
+            }
+            final int length = bases.length;
+            final boolean rc = rec.isReverseStrand();
+            ensureArraysBigEnough(length + 1);
+            if (rec.isPaired() && rec.isSecondOfPair()) {
+                seenSecondEnd = true;
+                for (int i = 0; i < length; i++) {
+                    final int cycle = rc ? length - i : i + 1;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: It looks like the zero element is never populated, which is slightly weird. I guess it's so you don't have to map index 0 to cycle one?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        boolean isEmpty() {
+            return maxLengthSoFar == 0;
+        }
+
+        public void addToMetricsFile(final MetricsFile<BaseDistributionByCycleMetrics, ?> metrics) {
+            int firstReadLength = 0;
+            for (int i = 0; i < maxLengthSoFar; i++) {
+                if (0 != firstReadCountsByCycle[i]) {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: reverse the arguments to if

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                    final BaseDistributionByCycleMetrics metric = new BaseDistributionByCycleMetrics();
+                    metric.READ_END = 1;
+                    metric.CYCLE = i;
+                    metric.PCT_A = (100.0 * firstReadTotalsByCycle[0][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_C = (100.0 * firstReadTotalsByCycle[1][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_G = (100.0 * firstReadTotalsByCycle[2][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_T = (100.0 * firstReadTotalsByCycle[3][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_N = (100.0 * firstReadTotalsByCycle[4][i] / firstReadCountsByCycle[i]);
+                    metrics.addMetric(metric);
+                    firstReadLength = i;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: This isn't really firstReadLength (it's 1 more than that), it's really firstReadCycles.

Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: or maybeMaxFirstReadCycles, since this is over many reads.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        final HistogramGenerator hist = filteredReads.aggregate(new HistogramGenerator(),
+                (hgp, read) -> hgp.addRead(read),
+                (hgp1, hgp2) -> hgp1.merge(hgp2));
+
+        final MetricsFile<BaseDistributionByCycleMetrics, Integer> metricsFile = getMetricsFile();
+        hist.addToMetricsFile(metricsFile);
+        return metricsFile;
+    }
+
+    private static Function<GATKRead, Boolean> makeReadFilter(final boolean pf_read_only, final boolean aligned_reads_only) {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: If this pattern becomes popular (as it appears to be doing), we should move it to a library. What do you think?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        return read -> (!pf_read_only || !read.failsVendorQualityCheck()) &&
+                (!aligned_reads_only || !read.isUnmapped()) &&
+                !read.isSecondaryAlignment() &&
+                !read.isSupplementaryAlignment();
+    }
+
+    protected void saveResults(final MetricsFile<?, Integer> metrics, final SAMFileHeader readsHeader, final String inputFileName) {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: This also seems very similar to the last one I reviewed, maybe it could be moved to a library.

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSparkIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public final class CollectBaseDistributionByCycleSparkIntegrationTest extends CommandLineProgramTest {
+    private static final File TEST_DATA_DIR = new File(getTestDataDir(), "picard/analysis/CollectBaseDistributionByCycle");
+
+    //Note: the 'expected' results in this test come from running picard 1.130
+    //NOTE: these tests use the same data and results as the non-spark ones, by design
```
Comment (davidadamsphd) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSparkIntegrationTest.java]: nit: Note

**Comment by davidadamsphd** [comment]
Overall, this looks fine, just a few minor comments.

Two broader questions/issues:
+ Does the test data exercise all of the code paths and edge cases?
+ In general, putting the majority of testing in integration tests instead of unit tests is bad pattern. It have several bad consequences (1) it becomes less clear which cases are being tested (2) it's slower than just running unit tests and (3) it makes it unhelpful to (perhaps someday) move to a testing framework that only runs tests relevant to the code directly affected (because all integration tests must be run).

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: or a superclass

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: it's this https://github.com/broadinstitute/gatk/issues/1026 right?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: i guess you're thinking about RNASeq. However, RNASeq data we have is done by reverse transcribing and sequencing the resulting DNA (called cDNA). About throwing - the metrics actually collects the number of 'unknown' bases so throwing would not work

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: agreed. renamed

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: agree it's weird. we need to preserve the behavior from picard though

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: ok

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: entered a ticket #1027

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 9 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: yup, i guess same #1027

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSparkIntegrationTest.java (hunk 10 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSparkIntegrationTest.java]: done

**Comment by akiezun** [comment]
1. line coverage is 89% for this class. Uncovered code is for a) read without bases, b)  IO exception when writing the report, c) plotting code. So i'm not too worried about this.

2. agreed in principle. To make this doable though we need a testing framework that would make it easy to manufacture reads en bulk. Right now we'd need to make them one by one - integration test has a nice property of being portable across picard/gatk and it's easy to plug in because it has an input with many reads.

I'm assuming that 'looks fine' is equivalent to 'LGTM' and will merge when tests pass.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 1 above)
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: Yup

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 3 above)
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: Sorry, what I meant to say is that unknown base is characterized by a specific character, right? If so, you should use that character directly. It seems strange to say that any other character is 'unknown' and not invalid.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java (hunk 5 above)
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java]: It'd be nice to note the lack of zero somewhere so it's doesn't seem like a bug to future readers.
